R/compengine.R

```diff
@@ -0,0 +1,815 @@
+#' http://physionet.org/physiotools/sampen/
+#' http://www.physionet.org/physiotools/sampen/matlab/1.1/sampenc.m
+#' Code by DK Lake (dlake@virginia.edu), JR Moorman and Cao Hanqing.
+#' 
+#' 
+#' @param y the input time series
+#' @param M embedding dimension
+#' @param r threshold
+#' 
+#' @references cf. "Physiological time-series analysis using approximate entropy and sample
+#' entropy", J. S. Richman and J. R. Moorman, Am. J. Physiol. Heart Circ.
+#' Physiol., 278(6) H2039 (2000)
+#' @references B.D. Fulcher and N.S. Jones. hctsa: A computational framework for automated time-series phenotyping using massive feature extraction. Cell Systems 5, 527 (2017).
+#' @references B.D. Fulcher, M.A. Little, N.S. Jones Highly comparative time-series analysis: the empirical structure of time series and their methods. J. Roy. Soc. Interface 10, 83 (2013).
+#' @author Yangzhuoran Yang
+#' @export
+sampenc <- function(y,M = 6,r = 0.3){
+  N <- length(y)
+  lastrun <- numeric(N) #zeros(1,N)
+  run <- numeric(N) #zeros(1,N)
+  A <- numeric(M) #zeros(M,1)
+  B <- numeric(M) #zeros(M,1)
+  # Get counting:
+  for(i in 1:(N-1)){ # go through each point in the time series, counting matches
+    y1 <- y[i]
+    for(jj in 1:(N-i)){ # compare to points through the rest of the time series
+      # Compare to future index, j:
+      j <- i + jj
+      # This future point, j, matches the time-series value at i:
+      if (abs(y[j]-y1) < r){
+        run[jj] <- lastrun[jj] + 1 # increase run count for this lag
+        M1 <- min(M, run[jj])
+        for (m in 1:M1){
+          A[m] <- A[m] + 1
+          if (j < N){
+            B[m] <- B[m] + 1
+          }
+        }
+      } else{
+        run[jj] <- 0
+      }
+    }
+    for( j in 1:N-i){
+      lastrun[j] <- run[j]
+    }
+  }
+  # Calculate for m <- 1
+  NN <- N*(N-1)/2
+  p <- A[1]/NN
+  e <- -log(p)  
+  return(e)
+}
+
+
+# stationarity ------------------------------------------------------------
+
+
+
+
+# SY_StdNthDer_1
+#' Standard deviation of the first derivative of the time series from software package \code{hctsa}
+#'
+#' Modified from \code{SY_StdNthDer} in \code{kctsa}. Based on an idea by Vladimir Vassilevsky.
+#' 
+#' @param y the input time series
+#' @return Standard deviation of the first derivative of the time series.
+#' @references cf. http://www.mathworks.de/matlabcentral/newsreader/view_thread/136539
+#' @references B.D. Fulcher and N.S. Jones. hctsa: A computational framework for automated time-series phenotyping using massive feature extraction. Cell Systems 5, 527 (2017).
+#' @references B.D. Fulcher, M.A. Little, N.S. Jones Highly comparative time-series analysis: the empirical structure of time series and their methods. J. Roy. Soc. Interface 10, 83 (2013).
+#' @author Yangzhuoran Yang
+#' @export
+Std1stDer <- function(y){
+  if(length(y)<2) stop("Time series is too short to compute differences")
+  yd <- diff(y)
+  return(sd(yd))
```

Comment:
What about missing values?

Reply:
I add a line next to it calculating the acf once
 `acfv <- stats::acf(x, length(x)-1, plot = FALSE)`
and then refer to the stored value using an argument whenever an acf call is required
`embed2_incircle_2 = embed2_incircle(x,2, acfv = acfv),`
so it won't repeatedly call the acf() function in each feature function.
Is this good enough? If not, could you give me some ideas?